### PR TITLE
Add functionality for dealing with JSON.

### PR DESCRIPTION
**Pull request by lrhn**

### CODE REVIEW

File: lib/src/json/reader/reader.dart
Comment (jonasfj) [inline lib/src/json/reader/reader.dart:35]: Why not call these `peak`?

File: lib/src/json/reader/reader.dart
Comment (jonasfj) [inline lib/src/json/reader/reader.dart]:
```suggestion
/// The JSON reader is intended to scan JSON source text from start to end.
```
(maybe, I imagine you're better at semantics)

File: lib/src/json/reader/reader.dart
Comment (jonasfj) [inline lib/src/json/reader/reader.dart:94]:
```suggestion
/// Calling methods out of order will result in unspecified behavior.
```

File: lib/src/json/sink/sink.dart
Comment (jonasfj) [inline lib/src/json/sink/sink.dart]: The API here isn't that bad. If the developer writes something like:
```dart
dynamic result;
final sink = jsonObjectWriter((r) => result = r);

// Write json to sink
writeSomeJsonToSink(sink);

// Do something with result
assert(result != null);
print(result);
```

But we risk that someone might do:
```dart
dynamic result;
final sink = jsonObjectWriter((result) {
  // Do something with result
  assert(result != null);
  print(result);  
});

// Write json to sink
writeSomeJsonToSink(sink);
```
Which could easily cause problems with try/catch, and other problems if `writeSomeJsonToSink` doesn't actually produce a result. At-least an unfortunate developer could easily create some messy code, if they don't employ a solid pattern here.

It might be wise to give consumers an example of which pattern you envision they employ when using this method (just in the doc comments).

You could also consider:
```dart
abstract class JsonSinkAndResult<T> extends JsonSink {
  /// May throw [StateError] if no value has been produced.
  /// This can probably depend on how much validation you have.
  T get result;
}
```
(but that creates another public type, I can see how that's not as elegant)

File: lib/src/json/sink/sink.dart
Comment (jonasfj) [inline lib/src/json/sink/sink.dart]: nit: Maybe use a static method like `JsonSink.stringWriter(StringSink sink, ...)`

(if I needed a `JsonSink` I might look at docs for the class to see if it has a constructor).

File: lib/src/json/builder/builder.dart
Comment (jonasfj) [inline lib/src/json/builder/builder.dart:58]: **Just a comment**: I see a lot of nice methods in library and wonder if somehow these could be static methods on some class... It seems unreasonable to introduce an empty abstract class for namespacing alone.. But when coding these will be hard to discover, and in documentation they might drown.

I guess `jsonOptional` and `jsonArray` could be extension methods on `Builder<T>`, but this also seems a bit crazy :)

I don't have a good suggestion for improving discoverability. I'm merely noting that this is a lot of top-level methods that pollutes docs and is hard to discover (I guess the upside is that it might be quite readable).

**Review by jonasfj — APPROVED** [review]
r+

**Comment by kevmoo** [comment]
Adding @jakemac53 – we should coordinate efforts on JSON here...

**Comment by kevmoo** [comment]
This feels like it should be its own package – at least to start. This is a lot of functionality we may want to evolve – without forcing breaking change releases on pkg:convert

**Review by kevmoo — CHANGES_REQUESTED** [review]
We should discuss this more  – maybe a VC – before we land!

File: lib/src/json/reader/reader.dart
Comment (lrhn) [reply lib/src/json/reader/reader.dart:35]: The `check` method returns true/false, not the value that a `peek` would traditionally do (return value if it's there, but without advancing).

The `try` functions (return value if it's there and *do* advance) are closer to `peek`, but I think more useful. Having both `check`, `try` and `peek` functions seemed like too much.

File: lib/src/json/reader/reader.dart
Comment (lrhn) [reply lib/src/json/reader/reader.dart]: It's not bad. My only worry is that the byte-reader doesn't read *text*, but then, the object-reader doesn't read *code* either. So, accepted.

File: lib/src/json/sink/sink.dart
Comment (lrhn) [reply lib/src/json/sink/sink.dart]: And I'm not a fan of the "ThingAndOtherThing" classes, and I do not want sinks to expose their final value.

I'd rather go for:
```dart
Object withJsonObjectWriter(void Function(JsonSink) action) { 
   Object result;
   bool hasResult = false;
   action(jsonObjectWriter((r) {
      result = r; 
      hasResult = true;
   }));
   if (!hasResult) throw StateError("Action did not write complete value");
   return result;
}
```
used as
```
int result = withJsonObjectWriter((sink) {
   sink.startArray();
   sink.writeInt(42);
   sink.writeInt(37);
   sink.endArray();
});
```

I guess I could parameterize it with a sink factory too, but then it gets really convoluted to use.

File: lib/src/json/sink/sink.dart
Comment (lrhn) [reply lib/src/json/sink/sink.dart]: I deliberately did not want to make a fixed set of writers. It's an abstract interface that anyone can implement, and these are just two examples.  Making them "first class" by putting factories on the interface is more privilege than I want to give them.

File: lib/src/json/builder/builder.dart
Comment (lrhn) [reply lib/src/json/builder/builder.dart:58]: The problem with putting them inside a class is that you need to write the class name to access them. And it's against the style guide to have a static-only class (and putting functions into an otherwise unrelated class is just a hack around that rule).

The traditional approach would be to make these top-level functions and import the library with a prefix. However, this is an existing library/package which is not imported with a prefix. So I added the `json` prefix to each function instead.

No perfect solution available. 

(And yes, extensions on a function type is crazy, and doesn't work for the generic variants because the function types are not generic).

**Comment by lrhn** [comment]
I can definitely make this a package by itself (that's how I wrote it, I just merged it into package:convert today).

As for Jake's work, that's mainly about serializing and deserializing Dart objects as JSON, not working with JSON by itself. That's a worthy goal too, but a different one.
This code allows you to work work directly with JSON data without having any corresponding Dart objects.

**Comment by jakemac53** [comment]
> As for Jake's work, that's mainly about serializing and deserializing Dart objects as JSON, not working with JSON by itself. That's a worthy goal too, but a different one.

Its both - but yes the primary focus is on the usability of parsing custom Dart classes from JSON.

These can be pretty intertwined though - a lot of what I have been looking at is in fact changing how we actually deal with JSON parsing (specifically on the web). A lot of performance gains I have seen come from changing how that is done.

Specifically, that largely comes down to using js interop to read individual keys from a JsObject on demand instead of eagerly creating a dart map/list. One of the biggest time savings there comes from not iterating all the keys of the JsObject, which isn't possible with this solution as far as I can tell. It always visits all keys so we would have to call `Object.getOwnPropertyNames` or some such (at least in package-land), which is expensive.

The difference here is basically between the visitor based approach which has to visit each key but can guarantee it only scans the bytes once, versus a model where you explicitly ask for individual keys that you need, but that would complicate and slow down the pure Dart implementation somewhat if we want to keep it lazy (you have to remember byte offsets for entries that you skipped, and re-scan them later on if somebody asks for the value).

I would like to see this be driven by some benchmarks, and with a web specific implementation as well.

Also some examples of what it looks like to deserialize into a custom object would be good.

**Comment by jakemac53** [comment]
Another sort of meta-comment is that this `JsonReader` class is pretty close to something that could be more generic and not specific to Json at all. If we could make it just an `ObjectReader` or something that would be a nice abstraction for general object serialization/deserialization (basically, its a `Decoder` from my previous explorations at that point).

**Comment by jakemac53** [comment]
One possibility here is an api where you ahead of time tell the parser about all the keys you care about for a reader - and then it would only visit the specified keys. This allows a simpler and faster pure dart implementation as well as a fast js interop based solution that doesn't require fetching all the keys on the js object.

File: lib/src/json/builder/builder.dart
Comment (jakemac53) [inline lib/src/json/builder/builder.dart:138]: It is likely common to not care about the keys you haven't specified. Should we have an additional boolean option here `skipUnspecifiedKeys` or some such?

File: lib/src/json/sink/sink.dart
Comment (jonasfj) [reply lib/src/json/sink/sink.dart]: However, `withJsonObjectWriter` is hard to use for sync and async constructions..

**Comment by jakemac53** [comment]
I do also think its worth discussing the merits of an event based parser versus a pull based parser like this.

Specifically when you consider the parsing of very large JSON blobs, it would be advantageous to be able to split the parsing up into multiple chunks and schedule them each on the event queue (and likely not the microtask queue). This allows the UI to render frames without jank during parsing of large JSON blobs, which is a known issue today.

With the current api you can achieve this, but it has to be built explicitly into the deserialization code the user writes for a given data structure (it could await an empty future between pulling values out of the reader). However, I don't think that is an ideal solution. You likely don't want to tie this decision to the data structure like that, but instead at a higher level where you are creating the JsonReader.

Essentially, you could just opt into using the `AsyncJsonReader` instead of the `JsonReader` at any point, but the code that actually does the deserialization into a given structure would look the same.

Note also that the reader itself can do a much better job of splitting up the work evenly and consistently than a user can.

File: lib/src/json/sink/sink.dart
Comment (lrhn) [reply lib/src/json/sink/sink.dart]: True.
The API here is inherently synchronous. A push-based API almost has to be because it has no way to recognize that an operation takes time.

I'd write a "buffering sink" if that becomes necessary, so you can buffer the events and deliver them when the sink is ready again.

File: lib/src/json/builder/builder.dart
Comment (lrhn) [reply lib/src/json/builder/builder.dart:138]: That's what a missing `defaultBuilder` will do.

**Comment by lrhn** [comment]
This is an API I wrote because I wanted to have it in situations where I was reading and writing JSON.
I am not interested in serializing or deserializing to Dart objects, but only directly working with the JSON programmatically. Creating intermediate objects *even using native `JSON.parse`* is something I explicitly want to avoid.

That does not mean that we cannot create a `JsonReader` backed by native JS values, like the result of JS `JSON.parse`. It's not something that can be done on all platforms, though.

This is a low-level API, as witnessed by it being so easy to use it incorrectly, like the requirement that arguments to `tryKey` must be sorted. It's definitely possible to use it incorrectly, and there is no attempt to eagerly catch that. It is not necessarily *ergonomic*. If you need to write helper methods to do some things, that's better than trying to do everything in the API itself.

Also, this is not intended to be a solution to all problems. I do not believe one exists, which is why we need *more* tools to choose between.

Providing a list of object key names, for example in `expectObject`, so that all other keys are ignored, is a viable idea. I can definitely add that. However, it requires more mutable state to store and restore when entering or leaving objects, whether you use the feature or not, and you can already get the effect using:
```dart
String tryKeyOrSkip(JsonReader reader, List<String> keys) {
   do {
     var key = reader.tryKey(keys);
     if (key != null) return key;
   } while (reader.skipObjectEntry());
   return null;
}
````
I don't want to add unnecessary operations to a low-level API, especially when they add a cost to code which doesn't use the feature.


As for "chunked parsing", I'm sure it's possible to make a JsonReader->JsonSink operation which can be paused at any time along the way.  It needs to reify its stack, instead of being recursive, but I see no problem with that. Since the reader is pull based and the sink is push based, you have the perfect position to control things when you are between the two.

**Comment by jakemac53** [comment]
There are a lot of nice benefits to this library - it handles validation and strong typing for nested collections nicely. That definitely does solve one of the core customer pain points that we have identified.

As an officially supported library we should make sure it carries its weight - is it significantly different enough from existing `jsonDecode` to justify its long term maintenance, code size cost, and any potential confusion it might cause our users to have competing supported products?

Does it fit well with our future plans for JSON, or does this mean we are likely to end up with 3 separate solutions?

Does it perform well (enough) on all our supported platforms, or are users in for a surprise when they try to run on a specific platform?

Is the intention that we would direct users at this package for JSON in our docs and do we have a plan for updating them? How do we message/justify the difference between these options in the docs? Do we deprecate existing `jsonDecode`?

If this is intended for personal use I would put it in a separate package under http://github.com/google/ (or personal repo), but as a supported package under dart-lang I think its important we take the time to answer questions like these before releasing it. And probably a fair number more people than are currently involved should sign off on it.

**Comment by kevmoo** [comment]
> If this is intended for personal use I would put it in a separate package under http://github.com/google/ (or personal repo), but as a supported package under dart-lang I think its important we take the time to answer questions like these before releasing it. And probably a fair number more people than are currently involved should sign off on it.

+💯 – let's evolve this model before we signup to support it forever w/ pkg:convert

**Comment by jakemac53** [comment]
Fwiw (sorry to keep spamming here) I did put together some basic benchmarks of my own - as well as taking a shot at a quick and dirty js interop based reader.

https://gist.github.com/jakemac53/97bad98f19feed59ae756ec30897a7a4

The tldr; is that the js-interop solution with JSON.parse is not very good (there might be a better implementation possible though).

However the native dart solution here (which would be preferable anyways) performs quite well even when compared with existing `jsonDecode` - for this particular benchmark it is about 25% faster to deserialize into my `Login` class in `-O3` mode which is arguably the most important benchmark.

That to me is a good signal that it is worth investing in this as a core supported thing.

**Comment by lrhn** [comment]
I'm almost ready to release this as an open source library using the usual open source process.

If we want it under dart-lang/ instead of my private github account, we can certainly do that too, then it will be a google supported project. We can also wait and promote it later if we want to.

**Comment by jakemac53** [comment]
Continuing forward with releasing under a private account SGTM. It might be good to circle back to it after NNBD is ready and look back into moving it under dart-lang (possibly package:json or something which we are squatting on I believe).

**Comment by lrhn** [comment]
@jakemac53 That's interesting benchmark results!

I'm considering whether `expectString`/`tryString` should also exist in a version which takes an existing list of possible values to avoid allocating a new string that we already know.

**Comment by jakemac53** [comment]
> @jakemac53 That's interesting benchmark results!

I think it would be very interesting to look into benchmarking a modified protobuf generator that uses this as well. cc @sigurdm just in case you felt like it :D

**Comment by jakemac53** [comment]
> As for "chunked parsing", I'm sure it's possible to make a JsonReader->JsonSink operation which can be paused at any time along the way. It needs to reify its stack, instead of being recursive, but I see no problem with that. Since the reader is pull based and the sink is push based, you have the perfect position to control things when you are between the two.

Fair enough - I think there is still a potential issue with very large individual entries but that is a somewhat special (and hopefully rare) case. Specialized parsers for apps that have to deal with that are probably not unreasonable (and json is already probably a bad choice for them).

File: lib/src/json/builder/builder.dart
Comment (natebosch) [inline lib/src/json/builder/builder.dart:18]: Can we consider `Object` (or `Object?` soon) for these APIs so that it isn't as easy to misuse the return value?

Comment (lrhn) [reply lib/src/json/builder/builder.dart:18]: Probably a good idea.
Dynamic is dangerous.

**Comment by jakemac53** [comment]
I went ahead and ported protobuf to the point where I could run the benchmarks (notably I didn't touch proto3).

Probably needs to be validated but I see significant wins on both the VM and Web - the exception being large string messages on the web which are slower (not surprising). Many benchmarks are 2x faster or more.

I went ahead and pushed two branches, one to just get the benchmarks running and default to `-O3` mode and dart2js [here](https://github.com/dart-lang/protobuf/tree/update-build-web-compilers). And another which also ports the `mergeFromJson` api to use this new api [here](https://github.com/dart-lang/protobuf/tree/json-parser-experiment).

cc @sigurdm 

Note that if you add an extra benchmark to the suite with larger strings then the difference becomes more noticeable. The break even point is somewhere around 30 characters, and at 1k characters it is close to 4x slower. I don't know if we have any idea how large string fields are on average though.

**Comment by lrhn** [comment]
This has now been released as a separate package [`jsontool`](https://pub.dev/packages/jsontool).

**Pull request closed without merging**
